### DISCUSSION TITLE
Using both Laravel Sanctum and Laravel Passport, together?

### DISCUSSION THREAD

**Discussion by lucasdcrk** [laravel/framework#42945] [Q&A]
Hi,

I searched online for a way to use both passport and sanctum, which is not possible at the time as both packages would override each other. Most of the answers says that there is no need for both of them, you either chooser the simplicity of Sanctum or the fully fledge passport.

But I think they should be able to be used complementarily, let's me explain:

I have an app with Sanctum, which I love for authenticating first party apps (made by me) and allowing my users to create API tokens in a simple manner.
But I would also like to provide OAuth for third party apps / utilities. Using sanctum is great for everything first party (users can enter they credentials on the mobile app safely) but I would rather use OAuth to avoid my users having to enter their password on an app made by a third party.

So I could switch sanctum and go for passport but I will lose the ability for my users to create API tokens simple and they would be forced to OAuth too (I guess?). Though I would still be able to use password grant for my mobile app.
Or I could keep sanctum and create my own wrapper for the PHP League OAuth package?

Has someone face the same challenge as me? Would love to hear your thoughts!

Regards,

**Comment by r0bdiabl0** [comment]
This would be great as they solve 2 entirely different use cases.

**Comment by schonhoff** [comment]
Hey,
I faced the same challenge and got it to work. I needed two User Models (one is the standard User and one is an ApiUser). After that you need some modifications on the auth.php config file:

```php
    'guards' => [
        'api' => [
            'driver' => 'passport',
            'provider' => 'users',
        ],

        'api-sanctum' => [
            'driver' => 'sanctum',
            'provider' => 'api-users',
        ],
        [...]
      ],

    'providers' => [
        'users' => [
            'driver' => 'eloquent',
            'model' => App\Models\User::class,
        ],

        'api-users' => [
            'driver' => 'eloquent',
            'model' => App\Models\ApiUser::class,
        ],

        [...]
    ],
```

Now the ApiUser can use Sanctums HasApiToken and the normal User can use Passports HasApiTokens. Both extends a BaseUser Model which holds the relations and such stuff.

Now you can use 'auth:api-sanctum' on routes as a middleware and an api-Passport middleware on other routes.

I found the solution by some googling, but I can't find it at the moment (maybe I'm using the wrong keywords again :D ). But maybe this is at least a direction you can work with.

**Reply by lucasdcrk** [reply]
Damn that’s smart, didn’t thought of doing this way 👌
Thanks!

**Reply by kyuchukovv** [reply]
@schonhoff What if I like to use both middlewares for same routes?

**Reply by schonhoff** [reply]
I would guess that this is not possible. But you can try it out locally

**Reply by kyuchukovv** [reply]
Currently I am trying yes, probably I don't need Passport. 
I am using Socialite to receive oauth2 providers data and save the tokens in db. Then I am logging the user and consuming api endpoints.
Will post if I succeed and is not a complete mess.

**Reply by avianheit07** [reply]
I've tried this approach but I keep getting 'Socket hang up' for unauthenticated users in sanctum. Have you encountered this? Can you provide more details on how to properly set this up please?

**Reply by schonhoff** [reply]
Sorry, but it works for us for 3 years now with the solution mentioned above. Maybe you did something wrong?

**Reply by jase-languasco** [reply]
Freaking brilliant!

**Comment by KateKoltsova** [comment]
But why can't you use passport tokens for all users?

**Reply by mrimran** [reply]
There can be number of scenarios, then common ones are: 
1 - The application is already built on sanctum and now that also required OAuth functionality.
2 - For standard auth process application wants to stay with simple token based approach of sactum and want to use passport for 3rd party auth systems.

**Comment by ismaildasci** [comment]
Hey! Yes, you can definitely use both Sanctum and Passport together. The trick is separate guards in config/auth.php:

```php
'guards' => [
    'sanctum' => [
        'driver' => 'sanctum',
        'provider' => 'users',
    ],
    'api' => [
        'driver' => 'passport',
        'provider' => 'users',
    ],
],
```

Then in routes, specify which guard:

```php
// First-party mobile app
Route::middleware('auth:sanctum')->group(function () {
    // sanctum protected routes
});

// Third-party OAuth clients
Route::middleware('auth:api')->group(function () {
    // passport protected routes
});
```

**Watch out for:**
1. Both packages have `HasApiTokens` trait - alias one if using both on User model
2. Token tables are separate (personal_access_tokens vs oauth_access_tokens)
3. Keep routes organized with different prefixes

For first-party mobile + third-party OAuth, this setup makes total sense!
